File: common/src/main/java/cn/breadnicecat/candycraftce/level/CaramelPortal.java
package cn.breadnicecat.candycraftce.level;

import cn.breadnicecat.candycraftce.block.CBlockTags;
import cn.breadnicecat.candycraftce.block.CBlocks;
import cn.breadnicecat.candycraftce.utils.Axes;
import cn.breadnicecat.candycraftce.utils.multiblocks.VectorPortalShape;
import net.minecraft.core.BlockPos;
import net.minecraft.world.level.Level;
import net.minecraft.world.level.block.state.BlockState;

import java.util.function.BiFunction;

import static cn.breadnicecat.candycraftce.block.CBlocks.CARAMEL_PORTAL;
import static cn.breadnicecat.candycraftce.block.blocks.CaramelPortalBlock.*;
import static net.minecraft.world.level.block.Blocks.LAVA;

/**
 * Created in 2025/2/4 15:22
 * Project: candycraftce
 *
 * @author <a href="https://github.com/Bread-Nicecat">Bread_NiceCat</a>
 * <p>
 *
 * <p>
 **/
public class CaramelPortal {
	public static final VectorPortalShape.PortalConfig CONFIG = new VectorPortalShape.PortalConfig(
			2, 21, 3, 21,
			true, true,
			b -> b.isAir() || /*b.is(CARAMEL_LIQUID.get()) ||*/ b.is(LAVA) || b.is(CARAMEL_PORTAL.get()),
			b -> b.is(CBlockTags.BT_CARAMEL_PORTAL_FRAME));
	
	public static final BiFunction<Axes, BlockState, BlockState> PLACER = (axes, old) -> {
		if (old.is(CARAMEL_PORTAL.get())) {
			return old.setValue(X, axes.hasX() | old.getValue(X))
					.setValue(Y, axes.hasY() | old.getValue(Y))
					.setValue(Z, axes.hasZ() | old.getValue(Z));
		} else return CARAMEL_PORTAL.defaultBlockState()
				.setValue(X, axes.hasX())
				.setValue(Y, axes.hasY())
				.setValue(Z, axes.hasZ());
	};
	
	public static void tryBuildOn(Level level, BlockPos pos) {
		VectorPortalShape.findPortalOnFrame(level, pos, CONFIG).ifPresent(t -> build(level, t));
	}
	
	/**
	 * 当前pos在框架内
	 */
	public static void tryBuildIn(Level level, BlockPos pos) {
		VectorPortalShape.findPortalInFrame(level, pos, CONFIG).ifPresent(t -> build(level, t));
	}
	
	public static void build(Level level, VectorPortalShape shape) {
		for (BlockPos frame : shape.getAllFrames()) {
			BlockState state = level.getBlockState(frame);
			if (state.is(CBlocks.SUGAR_BLOCK.get())) {
				level.setBlockAndUpdate(frame, CBlocks.CARAMEL_BLOCK.defaultBlockState());
			}
		}
		shape.build(level, PLACER);
	}
}

File: src/main/kotlin/cn/breadnicecat/candycraftce/multiblock/caramel_portal/CaramelPortal.kt
package cn.breadnicecat.candycraftce.multiblock.caramel_portal

import cn.breadnicecat.candycraftce.utils.AxisSet
import net.minecraft.core.BlockPos
import net.minecraft.core.Direction
import net.minecraft.world.level.Level
import java.util.stream.Stream

/**
 * Created by NiceCat on 2025/12/9.
 * Project: candycraftce
 * @author <a href="https://github.com/BreadNiceCat">Bread_NiceCat</a>
 *
 */
sealed interface CaramelPortal {
    fun getPortals(): Stream<BlockPos>
    fun getRequiredFrames(): Stream<BlockPos>
    fun getOptionalFrames(): Stream<BlockPos>
    fun getPortalsAxis(): Stream<Pair<BlockPos, AxisSet>>
    fun getAllFrames(): Stream<BlockPos> = Stream.concat(getRequiredFrames(), getOptionalFrames())
    fun getUnits(): Stream<Flat>
    fun validate(level: Level, config: PortalConfig): Boolean {
        for (pos in getRequiredFrames()) {
            if (!config.isFrame(level.getBlockState(pos))) return false
        }
        for (pos in getPortals()) {
            if (!config.isEmpty(level.getBlockState(pos))) return false
        }
        return true
    }

    fun build(level: Level, placer: PortalPlacer) {
        getPortalsAxis().forEach { (pos, ax) ->
            level.setBlock(pos, placer.place(ax, level.getBlockState(pos)), 3)
        }
    }

    override fun toString(): String
    override fun hashCode(): Int

    class Flat internal constructor(
        val base: BlockPos,
        val axis: Direction.Axis,
        val pipe2: List<Direction.Axis>,
        val width: Int,
        val height: Int,
        private val portals: Set<BlockPos>,
        private val requiredFrames: Set<BlockPos>,
        private val optionalFrames: Set<BlockPos>,
    ) : CaramelPortal {
        override fun getPortals(): Stream<BlockPos> = portals.stream()
        override fun getRequiredFrames(): Stream<BlockPos> = requiredFrames.stream()
        override fun getOptionalFrames(): Stream<BlockPos> = optionalFrames.stream()
        override fun getPortalsAxis(): Stream<Pair<BlockPos, AxisSet>> = portals.stream()
            .map { it to AxisSet(axis) }

        override fun getUnits(): Stream<Flat> = Stream.of(this)

        val end: BlockPos = base.relative(pipe2[0], width - 1).relative(pipe2[1], height - 1)

        override fun toString(): String = """
          {
            "type": "unit",
            "base": {
                "x": ${base.x},
                "y": ${base.y},
                "z": ${base.z}
            },
            "axis": "$axis",
            "width": $width,
            "height": $height
          }
        """.trimIndent()

        override fun hashCode(): Int {
            var result = width * height
            result = 31 * result + base.hashCode()
            result = 31 * result + axis.hashCode()
            return result
        }

        override fun equals(other: Any?): Boolean {
            if (this === other) return true
            return true
        }
    }

    class Compound internal constructor(val parts: Set<CaramelPortal>) : CaramelPortal {
        val requiredFrames: Set<BlockPos> by lazy {
            val required = LinkedHashSet<BlockPos>()
            parts.forEach {
                it.getRequiredFrames().forEach(required::add)
            }
            required
        }
        val optionalFrames: Set<BlockPos> by lazy {
            val optional = LinkedHashSet<BlockPos>()
            parts.forEach {
                it.getOptionalFrames().forEach(optional::add)
            }
            optional
        }
        val portals: Map<BlockPos, AxisSet> by lazy {
            val portals = LinkedHashMap<BlockPos, AxisSet.MutableAxisSet>()
            parts.forEach {
                if (it is Flat) {
                    it.getPortals().forEach { pos ->
                        if (pos in portals) {
                            portals[pos]!!.with(it.axis)
                        } else {
                            portals[pos] = AxisSet.MutableAxisSet(it.axis)
                        }
                    }
                } else if (it is Compound) {
                    it.portals.forEach { (pos, axs) ->
                        if (pos in portals) {
                            portals[pos]!!.withOr(axs)
                        } else {
                            portals[pos] = axs.mutable()
                        }
                    }
                }
            }
            portals
        }

        override fun getPortals(): Stream<BlockPos> = portals.keys.stream()
        override fun getRequiredFrames(): Stream<BlockPos> = requiredFrames.stream()
        override fun getOptionalFrames(): Stream<BlockPos> = optionalFrames.stream()
        override fun getPortalsAxis(): Stream<Pair<BlockPos, AxisSet>> =
            portals.entries.stream().map { it.toPair() }

        override fun getUnits(): Stream<Flat> {
            return parts.stream().flatMap { it.getUnits() }
        }

        override fun toString(): String = """
          {
            "type": "compound",
            "parts": [${parts.joinToString(", ")}]
          }
        """.trimIndent()

        override fun hashCode(): Int = parts.hashCode()
        override fun equals(other: Any?): Boolean {
            return this === other
        }

    }
}